R/count_teleconnections.R

```diff
@@ -149,21 +150,42 @@ count_watershed_teleconnections <- function(data_dir,
           .[["GCAM_ID"]] %>% unique() %>%
           length() -> tc_n_cropcover
 
-        # filter out where "is crop" is false and count land classes, adding 1 class to account for agriculture.
-        crop_and_landcover_types %>%
-          filter(is_crop == FALSE) %>%
-          .[["CDL_Class"]] %>% unique() %>%
-          length() -> number_landclasses_ex_ag
-
-        if (tc_n_cropcover > 0){
-          tc_n_landclasses = number_landclasses_ex_ag + 1
-        }else{
-          tc_n_landclasses = number_landclasses_ex_ag
-        }
         # TELECONNECTION - NUMBER OF FLOOD CONTROL DAMS WITHIN WATERSHED.
         flood_control_dams[watersheds_city, ] %>%
           length() -> tc_fcdam
 
+        # TELECONNECTION - RANK WATERSHED BASED ON % OF DEVELOPED/CULTIVATED AREA.
+        # Get frequencies of each type of land cover within the designated watershed.
+        mask_raster_to_polygon(cropcover_USA, watersheds_city) %>%
```

Comment:
This is right way to draw in the required information and I'm glad your using the pre-existing functions. However, we now have a slight problem in that the the masking of the raster layer is being performed twice: once for the cropper ids and once for this part. Since the crop masking is the biggest bottleneck in terms of speed, we only want to have to do it one time for each city. So I propose something like the following:
1. Update the function `get_raster_val_classes` so that it returns the matrix of raster values rather than just the classes. This can be done quite easily be changing the `unique()` call at the bottom of this function to `values` instead.
2. Since this function will no longer be returning the classes, you'll need to use a `unique()` to get the crop_ids (line ~140 in count_teleconnections).
3. Now, since you'll have this matrix of values as the output of `get_raster_val_classes()`, you don't need to mask the raster again. Instead you can simply use `table()` to turn the matrix of values into frequencies (and this drops NA automatically, so you'll need neither `freq()` nor the `is.na()` filter).

Reply:
@swd-turner doing it this way seems to return a table that doesn't show how many times each class occurs. The frequency column is either 1 or 0, 1 being that the class does occur within the area. I need the amount of class pixels that are present within the area so that I can divide that by the total number of pixels.

Reply:
Hey @KristianNelson ... If I use `table()` on a matrix it returns a named vector of frequencies - e.g.:
```
> matrix(c(1,1,2,2,5,5,4,51,6), 3, 3) %>% table()
.
 1  2  4  5  6 51 
 2  2  1  2  1  1
```
Can this not be done on the result of `values()` ?

Reply:
@KristianNelson note that `values()` is part of the raster package, so you'll need to load this library or use `raster::` . Here's the result I got for Portland | OR:
```
.
   111    121    122    123    131    141    142    143    152    176    190    195 
  5693   1496     15      7    191    265 279043   1643  23511     34     24     26 
```

Reply:
I substituted `unique()` for `values()` and could not get it to work for the crop covers raster. I have made some changes that clean it up a bit while also let it function properly. I'll push up and we can go back over and see what can get cleaned up more.